packages/apollo-server-core/src/requestPipeline.ts

```diff
@@ -359,10 +366,39 @@ export async function processGraphQLRequest<TContext>(
   }
 
   async function execute(
-    document: DocumentNode,
-    operationName: GraphQLRequest['operationName'],
-    variables: GraphQLRequest['variables'],
+    requestContext: WithRequired<
+      GraphQLRequestContext<TContext>,
+      'document' | 'operationName' | 'operation'
+    >,
   ): Promise<ExecutionResult> {
+    const executors: GraphQLExecutor<TContext>[] = [];
+    if (config.executor) {
+      executors.push(config.executor);
+    }
+    dispatcher.requestListeners.forEach(requestListener => {
+      if (requestListener.executor) {
+        executors.push(requestListener.executor.bind(requestListener));
+      }
+    });
+    executors.push(defaultExecutor);
+
+    for (const executor of executors) {
```

Comment:
I was hoping we could extract this into something like `Dispatcher.invokeHookSequentially`, but I realize what's keeping that from working is the ability to pass in an `executor` outside of a plugin (which we want for the gateway).

To be honest, using the idea of an `executor` to return a response from cache still feels out of place to me. We always invoke `executionDidStart`/`executionDidEnd` for example, which I don't think we should do for caches. And we may want to rethink other hooks as well, based on whether an execution result or a preformatted response was returned. Right now for example, executors are defined as returning an `ExecutionResult`, which contains an `errors` property with `GraphQLError` instances. Those will need to be fed through `formatError` to be serializable. A response from cache on the other hand, will contain `GraphQLFormattedError`s.

An alternative might be to separate the notion of an `executor` from the ability to short-circuit execution through something like `responseForOperation` (which we would use for caching). We could then define `executor ` as `config.executor || defaultExecutor`, similar to how the `fieldResolver` property is used in `graphql-js`.

Reply:
Ok, you've said this like 5 times now so perhaps I will listen. In that case I'll just back out the executor change entirely and let it be its own PR.

Reply:
Or really, that means just dropping this PR and rebasing the cache PR on master.